File: rpsls_j17/src/main/java/nl/bentels/rpsls/Game.java
package nl.bentels.rpsls;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.StringReader;
import java.util.Scanner;

public class Game {
    private final PlayerRole ourRole;

    public Game() {
        this(PlayerRoleFactory.random());
    }

    Game(PlayerRole ourRole) {
        this.ourRole = ourRole;
    }

    public void doGame() {
        boolean chosen = false;
        do {
            try {
                System.out.print("Which role do you want (ROCK, PAPER, SCISSORS, LIZARD, SPOCK)? ");
                Scanner scanner = new Scanner(System.in);
                PlayerRole chosenRole = PlayerRoleFactory.fromName(scanner.nextLine());
                chosen = true;
                System.out.println("You chose %s".formatted(chosenRole.getClass().getSimpleName()));
                System.out.println("I chose %s".formatted(ourRole.getClass().getSimpleName()));
                System.out.println(
                        switch (chosenRole.match(ourRole)) {
                            case WIN -> "You win!!";
                            case LOSE -> "You lose..";
                            case DRAW -> "It's a draw";
                        }
                );
            } catch (IllegalArgumentException iae) {
                System.out.println(iae.getMessage());
                System.out.println("Try again");
            }
        } while (!chosen);
    }
}

File: rpsls_mvn/src/main/kotlin/Game.kt
class Game () {
    internal var _myRole: PlayerRole? = null
    val myRole
        get() = _myRole ?: PlayerRole.random()

    constructor(forcedRole: PlayerRole) : this() {
        _myRole = forcedRole
    }

    fun playGame(): Unit {
        var chosen = false
        while (!chosen) {
            try {
                print("Which role do you want (${PlayerRole.values().joinToString()})? ")
                val choice = readLine()!!
                val chosenRole = PlayerRole.valueOf(choice.uppercase())
                chosen = true
                val result = chosenRole.compareWith(myRole)
                println("You chose ${choice.uppercase()}")
                println("I chose ${myRole}")
                println(
                    when (result) {
                        RoleComparisonResult.WIN, RoleComparisonResult.LOSE -> "You ${result}"
                        RoleComparisonResult.DRAW -> "It's a ${result}"
                    }
                )
            } catch (nsee: NoSuchElementException) {
                println("That's not a role!! Try again...")
            }
        }
    }
}